`RecommendedCutoffMixin`: set only one stringency with `set_cutoffs`

mbercx opened the pull request:
Fixes #68 

Currently the behaviour of the `set_cutoffs` and `get_cutoffs` methods of the
`RecommendedCutoffMixin` class is different. The first sets the cutoffs for all
stringencies along with a default stringency, the second only retrieves the
cutoffs for a specified stringency. This can be somewhat confusing for the
user, and means that the user cannot easily set the cutoffs for a single
stringency.

Here we change the `set_cutoffs` method so it only sets the recommended
cutoffs for a single stringency, which has to be specified as an input
argument. The stringency is automatically set to be the default in case
it is the only stringency specified, and a `set_default_stringency`
method is added to the user can still change the default stringency at
any time after setting multiple stringencies.

We also allow the user to specify units for each stringency, extending
the value corresponding to the `_key_cutoffs_unit` in the extras to a
dictionary that specifies the units for each stringency, similar to the
cutoffs.

mbercx commented:
@sphuber test should be passing, but could still do with a good scrubbin'. Will do so once we are agreed on the design!

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -90,60 +91,92 @@ def validate_stringency(self, stringency: str) -> None:
         """
         return self.get_extra(self._key_cutoffs, {})
 
+    def _get_cutoffs_unit_dict(self) -> str:
```
```suggestion
    def _get_cutoffs_unit_dict(self) -> dict:
```

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -90,60 +91,92 @@ def validate_stringency(self, stringency: str) -> None:
 
         :return: the default stringency or ``None`` if no cutoffs have been defined for this family.
-        :raises ValueError: if not default stringency has been defined.
+        :raises ValueError: if default stringency has not been defined.
```
I noticed that the current docstring and return type of `get_default_stringency` is incorrect. It says it returns `None` if not defined but it actually raises. Maybe raising is actually the best behavior? If we have it return `None`, we should maybe change the implementation of `get_cutoffs_unit` which now assumes it gets a value back

mbercx replied:
I guess it depends on what error we want the user to see when they try to use `get` methods and have no stringencies defined. Once _any_ stringency is defined with the `set_cutoffs` method, it will automatically be set to the default. I think raising with an error message like `'no recommended cutoffs have been specified for this pseudo family'` might be better than `'no default stringency has been defined.'` when no stringencies have been defined _at all_.

However, now that I think of it, we should probably also add a `remove_cutoffs` method, since now there is no API for doing so. If this method is implemented, it's possible that the user:

1. Uses `set_cutoffs` to set the `'normal'` stringency, which becomes the default.
2. Uses `set_cutoffs` to set to `'high'` stringency.
3. Removes the `'normal'` stringency.

What do we do in this instance? Since there is only one stringency left, it makes sense to make that the default, but what if there are 3 stringencies at some point, and the default is removed? Do we let the user simply not remove the default stringency (raise), or allow stringencies/cutoffs to be defined without having a default set?

sphuber replied:
Good points. Seems like that really we should just accept that it is very much possible to not have a default stringency defined, which means that we should simply have it raise when it is asked for. I think that `'no default stringency has been defined'` in this case is then better than `'no recommended cutoffs have been specified for this pseudo family' ` because not only does the latter tie an implementation detail into error message, but once we add the `delete_cutoffs` method, that won't necessarily be true anymore.

TLDR: my vote
* let `get_default_stringency` raise if not defined
* `set_cutoffs` automatically sets default if only one stringency is defined
* implement `delete_cutoffs` that will also remove the default stringency if that corresponds (only sets new default if only stringency remains)

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -90,60 +91,92 @@ def validate_stringency(self, stringency: str) -> None:
+    def set_default_stringency(self, default_stringency: str) -> None:
+        """Set the default stringency for the recommended cutoffs.
+
+        :param default_stringency: the default stringency to be used when ``get_recommended_cutoffs`` is called. If is
```
```suggestion
        :param default_stringency: the default stringency to be used when ``get_recommended_cutoffs`` is called. It is
```

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -90,60 +91,92 @@ def validate_stringency(self, stringency: str) -> None:
+        """Set the default stringency for the recommended cutoffs.
+
+        :param default_stringency: the default stringency to be used when ``get_recommended_cutoffs`` is called. If is
+            possible to not specify this if and only if the cutoffs only contain a single stringency set. That one will
+            then automatically be set as default.
```
I think these last two sentences of the docstring actually apply to `set_cutoffs` doesn't it? Here it doesn't make much sense.

mbercx replied:
Oops! Copy 🍝 ! 😬

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -90,60 +91,92 @@ def validate_stringency(self, stringency: str) -> None:
+            the pseudo family.
+        """
+        if default_stringency not in self.get_cutoff_stringencies():
+            raise ValueError('provided default stringency not in tuple of available cutoff stringencies.')
```
```suggestion
            raise ValueError(f'provided default stringency not in tuple of available cutoff stringencies: {self.get_cutoff_stringencies}.')
```

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -90,60 +91,92 @@ def validate_stringency(self, stringency: str) -> None:
+        """Set the recommended cutoffs for the pseudos in this family and a specified stringency.
 
-        .. note:: units of the cutoffs should be in electronvolt.
+        .. note: If there is only one stringency defined for the pseudo family, this method will automatically set this
+            as the default. Use the ``set_default_stringency`` method to change the default when setting multiple
+            stringencies.
```
```suggestion
        .. note: If, after the cutoffs have been set, there is only one stringency defined for the pseudo family, this method will automatically set this
            as the default. Use the ``set_default_stringency`` method to change the default when setting multiple
            stringencies.
```

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -155,7 +188,19 @@ def get_cutoffs(self, stringency=None) -> Union[dict, None]:
 
         try:
-            return self._get_cutoffs()[stringency]
+            return self._get_cutoffs_dict()[stringency]
```
Should probably add to the docstring:
```
:raises ValueError: if no stringency is specified and no default stringency is defined for the family.
```

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -155,7 +188,19 @@ def get_cutoffs(self, stringency=None) -> Union[dict, None]:
+            raise ValueError(f'stringency `{stringency}` is not defined for this family.') from exception
+
+    def get_cutoffs_unit(self, stringency: str = None) -> str:
+        """Return the cutoffs unit.
```
```suggestion
        """Return the cutoffs unit for the specified or family default stringency.
```

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -155,7 +188,19 @@ def get_cutoffs(self, stringency=None) -> Union[dict, None]:
+
+    def get_cutoffs_unit(self, stringency: str = None) -> str:
+        """Return the cutoffs unit.
+
```
```suggestion
:param stringency: optional stringency for which to retrieve the unit. If not specified, the default stringency of the family is uses.
```

sphuber commented on tests/cli/test_family.py:
```diff
@@ -13,32 +13,49 @@
 
 @pytest.fixture
-def generate_cutoffs(tmp_path):
+def generate_cutoffs_dict(tmp_path):
```
```suggestion
def generate_cutoffs_dict():
```

sphuber replied:
You can also have this fixture reuse the `generate_cutoffs` fixture perhaps to generate the inner structure?

sphuber commented on tests/cli/test_family.py:
```diff
@@ -13,32 +13,49 @@
+
+
+@pytest.fixture
+def generate_cutoffs(tmp_path):
```
```suggestion
def generate_cutoffs():
```

sphuber commented on tests/cli/test_family.py:
```diff
@@ -57,28 +74,28 @@ def test_family_cutoffs_set(run_cli_command, get_pseudo_family, generate_cutoffs
     # unit can be used at a time. This limitation will soon be removed though, at which point, this family should have
     # eV as a unit, such that the test of the CLI call to set it to hartree actually tests that the unit is changed.
-    family.set_cutoffs(cutoffs, unit='hartree', default_stringency=stringency)
+    family.set_cutoffs(cutoffs, stringency, 'hartree')
```
```suggestion
    family.set_cutoffs(cutoffs, stringency, 'eV')
```
use an explicitly different unit than the one we set later for the actual test. This is now possible because we allow multiple units to be specified. You can therefore also remove the three comments above this, or replace them with an explanation why we set `eV` explicitly (this is just in case in the future for some reason the default unit is changed to `hartree`)

mbercx replied:
Ah, I missed that comment while swashbuckling through the tests somehow. My bad!

sphuber reviewed (changes requested):
Thanks @mbercx some minor adjustments but I think the design of the interface is good like this. Maybe just add some tests surrounding the fact that we now have multiple units and add a test on what should happen if you call it and no default stringency is set.

sphuber commented on aiida_pseudo/groups/mixins/cutoffs.py:
```diff
@@ -90,72 +89,143 @@ def validate_stringency(self, stringency: str) -> None:
+        except KeyError as exception:
+            raise ValueError(f'stringency `{stringency}` is not defined for this family.') from exception
+
+    def delete_cutoffs(self, stringency) -> None:
```
![oliver](https://user-images.githubusercontent.com/6992332/114423667-b13fc900-9bb7-11eb-9992-98fb540247ab.png)

And maybe we do the same as in `set_cutoffs`, if after the operation is completed there is only one stringency left, that becomes the new default. If you agree, would be good to have a test for this and add in the docstring like for `set_cutoffs`

mbercx replied:
I agree it makes sense, and don't immediately see any issue with this. I wonder if we should print a message when the user removes the `default_stringency` with the `remove_cutoffs` method. Something along the lines of:

* Only one stringency remaining: `The default stringency has been set to: '<LEFTOVER STRINGENCY>'`
* Multiple stringencies remaining: `Default stringency removed but multiple options left. Please set the new default stringency with the ``set_default_stringency()`` method.`

And I will make sure that Oliver's hunger for tests is amply satisfied. 😁

sphuber replied:
Would be fine with me, but what are you suggesting to use for printing the message? I think the only options we have are logging or a `warnings.warn` and probably the latter since we haven't really used or setup logging yet. Printing directly I don't like in normal API code tbh.

mbercx commented:
That should turn Oliver's frown upside down! I also implemented the logic discussed and added some warnings in case the default stringency is removed, or a new default stringency needs to still be set.

sphuber reviewed (approved):
"Do you know what I consider the greatest sin in the world, my dear? Ingratitude." But I, for one, am very grateful for this PR
